app/domain/authentication/authn_oidc/authenticator.rb

```diff
@@ -0,0 +1,84 @@
+module Authentication
+  module AuthnOidc
+    class AuthenticationError < RuntimeError; end
+    class NotFoundError < RuntimeError; end
+    class OIDCConfigurationError < RuntimeError; end
+    class OIDCAuthenticationError < RuntimeError; end
+
+    class Authenticator
+
+      def initialize(env:)
+        @env = env
+      end
+
+      def valid?(input)
+        @authenticator_name = input.authenticator_name
+        @service_id = input.service_id
+        @conjur_account = input.account
+        @request_body = input.request.body.read
```

Comment:
The `valid?` method should be pure: It shouldn't be mutating state on the object.

This is problematic for multiple reasons:
1. It makes the code flow more complex to reason about.
2. (see my other comment on `CommandClass`) These authenticator objects are ultimately intended to be created once, at boot time.  When we actually do this (which will probably be soon), code like the code above will actually not be threadsafe.

The `command_class` gem offers a simple alternative that will make this kind of code unnecessary.

In the case of `request_body` you'd simply create a memoized private method (`@input` will be automatically available):
```
def request_body
  @request_body ||= @input.request.body.read
end
```

For the other 3, you could either do the same, or use the `Forwardable` class since this is just a case of delegation to `@input` .  For an example see:

https://github.com/cyberark/conjur/blob/master/app/domain/authentication/authn_k8s/validate_pod_request.rb#L17

Reply:
I added a ticket for this, we'll get to it after the split

Reply:
This one really needs to be fixed before we integrate this code into master.  If there's a time issue I can take this particular refactor off your hands since I'll be able to implement it quickly I believe.